typing/typedecl.ml

```diff
@@ -756,7 +756,22 @@ let check_well_founded env loc path to_check ty =
           let ty0 = if TypeSet.is_empty parents then ty else ty0 in
           check ty0 (TypeSet.add ty parents) ty'
         with
-          Ctype.Cannot_expand -> may raise arg_exn
+          Ctype.Cannot_expand ->
+            may raise arg_exn;
+            (* MPR#7726: Recursive modules allow invalid recursive types *)
+            if recmod && to_check p &&
+              let rec is_virtual_functor_app ~app = function
+                | Path.Pident _ as p  -> app && Env.is_functor_arg p env
+                | Path.Pdot (p, _, _) -> is_virtual_functor_app ~app p
+                | Path.Papply (p1, _) -> is_virtual_functor_app ~app:true p1
```

Comment:
Shouldn't we also look into the rhs of Papply (leaving ~app unchanged)?  I could not build a counter-example where this would be required, but it is not obvious why this is not needed.

Reply:
The point is that if `p1` is not a "virtual functor", i.e. if it has a concrete definition, then either it defines the type we are checking as an abbreviation, which will be expanded and checked again, or it defines a non-expandable type, which is contractive by definition.

I'm not sure it makes a big difference in terms of interesting accepted programs, but no need to be more restrictive in this case.